automap/src/comm_layer/igdp.rs

```diff
@@ -0,0 +1,1687 @@
+        tlh.exists_log_containing("ERROR: IgdpTransactor: Change handler recovery failed: discarded Error(NoLocalIpAddress)");
+    }
+
+    #[test]
+    #[should_panic(expected = "No JoinHandle: can't stop housekeeping thread")]
+    fn stop_housekeeping_thread_handles_missing_join_handle() {
+        let mut subject = IgdpTransactor::new();
+        let (tx, _rx) = unbounded();
+        subject.inner_arc.lock().unwrap().housekeeping_commander_opt = Some(tx);
+        subject.join_handle_opt = None;
+
+        let _ = subject.stop_housekeeping_thread();
+    }
+
+    #[test]
+    fn stop_housekeeping_thread_handles_panicked_housekeeping_thread() {
+        init_test_logging();
+        let mut subject = IgdpTransactor::new();
+        let (tx, _rx) = unbounded();
+        subject.inner_arc.lock().unwrap().housekeeping_commander_opt = Some(tx);
+        subject.join_handle_opt = Some(thread::spawn(|| panic!("Booga!")));
+
+        let change_handler = subject.stop_housekeeping_thread();
+
+        change_handler(AutomapChange::Error(AutomapError::CantFindDefaultGateway));
+        let tlh = TestLogHandler::new();
+        tlh.exists_log_containing(
+            "WARN: IgdpTransactor: Tried to stop housekeeping thread that had panicked",
+        );
+        tlh.exists_log_containing("ERROR: IgdpTransactor: Change handler recovery failed: discarded Error(CantFindDefaultGateway)");
+    }
+
+    #[test]
+    fn thread_guts_does_not_remap_if_interval_does_not_run_out() {
+        init_test_logging();
+        let (tx, rx) = unbounded();
+        let change_handler: ChangeHandler = Box::new(move |_| {});
+        let gateway = GatewayWrapperMock::new()
+            .get_external_ip_result(Ok(Ipv4Addr::from_str("1.2.3.4").unwrap()));
+        // No call to add_port_result; if the MUT tries to remap a port, the test will fail.
+        let inner_arc = Arc::new(Mutex::new(IgdpTransactorInner {
+            gateway_opt: Some(Box::new(gateway)),
+            housekeeping_commander_opt: Some(tx.clone()),
+            public_ip_opt: None,
+            mapping_adder: Box::new(MappingAdderMock::new()), // no provision for add_mapping()
+            logger: Logger::new("no_remap_test"),
+        }));
+        tx.send(HousekeepingThreadCommand::InitializeMappingConfig(
+            MappingConfig {
+                hole_port: 6666,
+                next_lifetime: Duration::from_secs(0),
+                remap_interval: Duration::from_secs(1),
+            },
+        ))
+        .unwrap();
+        tx.send(HousekeepingThreadCommand::Stop).unwrap();
+
+        let _ = IgdpTransactor::thread_guts(10, change_handler, inner_arc, rx);
+
+        // If we get here, neither mapping_adder.add_mapping() nor gateway.add_port() was called
+        TestLogHandler::new().exists_no_log_containing("INFO: no_remap_test: Remapping port 1234");
```

Comment:
Not sure, shouldn't this be "..: Remapping port 6666" rather?
These negative assertions are awful, I learn from you to know what you mean by talking similar.